### PR TITLE
Add test to PVForecast

### PR DESCRIPTION
**Pull request by b0661**
Tests were added for PVForecast

To improve testability several steps were done. Coverage currently is around 85%.

- Documentation of the module
- Addition of type hints where possible
- Separation, test and use of common utility functions to allow for general testing at one spot.
   - to_datetime
   - CacheFileStore
- Use of logging instead of print to easily capture in testing.
- Add validation of the json schema for Akkudoktor PV forecast data.
- Allow to create an empty PVForecast instance as base instance for testing.
- Make process_data() complete for filling a PVForecast instance for testing.
- Normalize forecast datetime to timezone of system given in loaded data.
- Do not print forecast data report but provide report for test checks.
- Get rid of cache file path using the CachFileStore to automate cache file usage.

### CODE REVIEW

**Comment by Lasall** [comment]
When we change to fastapi, we will use pydantic. Maybe we could replace the jsonschema with pydantic here.

**Comment by b0661** [comment]
> When we change to fastapi, we will use pydantic. Maybe we could replace the jsonschema with pydantic here.

Switched to pydantic.

**Review by Lasall — COMMENTED** [review]
I think currently only the url is used to fill PVForecast with data (/pvforecast). So to me the interface seems complicated supporting data (dict), filepath and url. If we want to support filepath and raw data input to the class maybe we should abstract filepath, url into PVForecastFile, PVForecastURL to do the io/network stuff (maybe one more abstraction layer for the caching) and then have the base class with all the logic operating directly on the raw data (e.g. process_data does not require any input, that's not already defined in base constructor).

File: src/akkudoktoreos/class_pv_forecast.py [hunk 1]
```diff
@@ -20,61 +129,165 @@ def __init__(
...
+            prediction_hours (Optional[int]): The number of hours to forecast into the future. Defaults to 48 hours.
+
+        Example:
+            forecast = PVForecast(data=my_forecast_data, start_date="2024-10-13", prediction_hours=72)
+        """
         self.meta = {}
         self.forecast_data = []
-        self.cache_dir = cache_dir
-        self.prediction_hours = prediction_hours
+        self.cache_files = CacheFileStore()
```
[the remark this answers is not in the document]
Comment (bernhard-tuvsud) [reply src/akkudoktoreos/class_pv_forecast.py:254]: I don´t know lru_cache but as far as I can see it is on function basis and caches in memory. Here caching is about caching some data retrieved from an URL up to a certain time in files. Usually for one day, maybe longer. Thats why you have to give a "valid until" date to build up the cache file.

Do you know an example how to do that with lru_cache?

**Comment by bernhard-tuvsud** [comment]
> I think currently only the url is used to fill PVForecast with data (/pvforecast). So to me the interface seems complicated supporting data (dict), filepath and url. If we want to support filepath and raw data input to the class maybe we should abstract filepath, url into PVForecastFile, PVForecastURL to do the io/network stuff (maybe one more abstraction layer for the caching) and then have the base class with all the logic operating directly on the raw data (e.g. process_data does not require any input, that's not already defined in base constructor).

You are right. I was thinking of creating a forecast class for every possible forecast source/ format (e.g. one for Akkudoktor.net - the current URL one) . Such a class could provide retrieving the raw forecast from URL, file, raw data. ATM this would radically change the interface. So better do it in another PR.

File: src/akkudoktoreos/class_pv_forecast.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (b0661) [reply src/akkudoktoreos/class_pv_forecast.py:254]: What about [cachetools](https://cachetools.readthedocs.io/en/latest/) TTLCache? Seems to be what is intended in the current code.

**Comment by b0661** [comment]
> I think currently only the url is used to fill PVForecast with data (/pvforecast). So to me the interface seems complicated supporting data (dict), filepath and url. If we want to support filepath and raw data input to the class maybe we should abstract filepath, url into PVForecastFile, PVForecastURL to do the io/network stuff (maybe one more abstraction layer for the caching) and then have the base class with all the logic operating directly on the raw data (e.g. process_data does not require any input, that's not already defined in base constructor).

You are right. Doing this would radically change the interface. I would do that in a following PR if this one is accepted.

File: src/akkudoktoreos/class_pv_forecast.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (b0661) [reply src/akkudoktoreos/class_pv_forecast.py]: Looked at [cachetools](https://cachetools.readthedocs.io/en/latest/) and [filecache](https://docs.sweep.dev/blogs/file-cache) and some other. They were missing some aspects I wanted to have. Will stay with CacheFileStore but added a function decorator `cache_in_file()` to make it quite easy to cache function return values.

```python
    @cache_in_file(mode="wb+")
    def load_data_from_url_with_caching(self, url: str) -> dict:
        response = requests.get(url)
        data = response.json()
        logger.debug(f"Data fetched from URL `{url} and cached.")
        return data

data = pv_forecast.load_data_from_url_with_caching(url, with_ttl="2 days")


```

**Comment by drbacke** [comment]
Please re-run all tests, code changed quite a lot

**Comment by drbacke** [comment]
Please re-run all tests, code changed quite a lot

**Comment by b0661** [comment]
> Please re-run all tests, code changed quite a lot

@drbacke, tests passed.

**Comment by b0661** [comment]
Tests fail due to unrelated bug #185 .

**Comment by b0661** [comment]
Tests fail due to unrelated bug #185 .

**Comment by b0661** [comment]
Test now pass again. Unrelated bug #185 really seems to be related to daylight saving time change yesterday.

**Pull request merged**